src/sources/util/unix.rs

```diff
@@ -19,3 +33,171 @@ pub fn change_socket_permissions(path: &Path, perms: Option<u32>) -> crate::Resu
+        Self {
+            uid: other.uid(),
+            gid: other.gid(),
+            pid: other.pid().clone(),
+        }
+    }
+}
+
+/// This structure defines the various kinds of metadata we can
+/// collect off a connected unix-domain socket and expose as source fields.
+#[derive(Clone, Debug)]
+pub struct UnixSocketMetadata {
+    /// The peer address of the socket, as returned from getpeername(2). This
+    /// will usually not be set (unless the connecting peer has explicitly
+    /// bound their socket to a path).
+    pub peer_path: Option<String>,
+
+    /// The inode/device of the socket, as collected from fstat(2). This only
+    /// makes sense for stream sockets, not datagram ones.
+    pub socket_inode: Option<SocketInode>,
+
+    /// The peer credentials of the process which connected to the socket,
+    /// reported by getsockopt(2) SO_PEERCRED/LOCAL_PEERCRED/etc. This may not
+    /// be the same as the process currently writing to the socket, because
+    /// the socket could have been passed down to a child process or sent
+    /// to a different process.
+    pub peer_creds: Option<SocketCredentials>,
+}
+
+impl UnixSocketMetadata {
+    pub fn peer_path_or_default(&self) -> &str {
+        match &self.peer_path {
+            Some(path) => &path,
+            None => UNNAMED_SOCKET_HOST,
+        }
+    }
+}
+
+
+/// Collects the device & inode number for a socket.
+pub async fn get_socket_inode<T : AsRawFd>(socket: &T) -> Result<SocketInode, Box<dyn std::error::Error>> {
+    // Get the socket file descriptor. This is the actual integer in use by the socket,
+    // not a dup(2) of it.
+    let socket_fd = socket.as_raw_fd();
+
+    // This needs to be done in a task, because fstat(2) is (technically) blocking.
+    // Tokio's file::metadata() essentially does the same thing.
+    spawn_blocking(move || {
+        // Construct a new std::file from it.
+        // We _really_ don't want std_file to run its drop (which would close the actual file!)
+        // under _any_ circumstances, because that's going to actually close the actual socket.
+        //
+        // Safety: from_raw_fd isn't really memory-unsafe, it's just warning us that
+        // double-closing of the descriptor might happen if it's still owned elsewhere.
+        // NonClosingFile fixes that.
+        let non_closing_socket_file = unsafe { NonClosingFile::from_raw_fd(socket_fd) };
+        non_closing_socket_file.file.metadata()
+    }).await
+        .map_err(|error| -> Box<dyn std::error::Error> {
+            // Propagate panics from the task
+            match error.try_into_panic() {
+                Ok(panic_reason) => resume_unwind(panic_reason),
+                // Err(error) => error.into(),
+                Err(error) => error.into(),
+            }
+        })
+        .and_then(|metadata_result| {
+            metadata_result.map_err(|error| -> Box<dyn std::error::Error> { error.into() })
+        })
```

Comment:
Yeah, this should indeed just be reworked to to use `tokio::fs::File` so that we can avoid the `spawn_blocking`/panic unwrapping stuff here. It's just extra noise.